### ISSUE TITLE
Heap Memory Consumption by using the TransactionalMap and IMap

### ISSUE THREAD

**Issue by vijaymaddi** [hazelcast/hazelcast#13572]
Hi Dev Team,

Testing **Hazelcast 3.8** in cloud oracle server  appears to be leaking heap memory in TransactionalMap and IMap .We are using **TransactionalMap** and **IMap** to store the data ,the heap memory keeps increasing even when we cleared and destroyed the Maps.We have checked the Map Statistics of Hazelcast  which shows heap memory as zero , backupmemory as zero, we have attached with **jprofiler** as well to see memory , instance Of TreeMap , HashMap keeps on increasing.When we shutdown the hazelcast the heap memory comes down to original heap size as before starting.

**Comment by mmedenjak** [comment]
Hi @vijaymaddi !

Can you post some heap dump here and/or add some code sample that would show us what was consuming the heap?

**Comment by mmedenjak** [comment]
@vijaymaddi any updates on the issue?

**Comment by mmedenjak** [comment]
Can you take a heap dump so we see what is consuming the heap?

**Comment by vijaymaddi** [comment]
okay , we will take heap dump and send to you soon

**Comment by vijaymaddi** [comment]
@mmedenjak ,
  I am attaching heap dump which have both formats csv and html

[heapdump.zip](https://github.com/hazelcast/hazelcast/files/2326863/heapdump.zip)

.

**Comment by mmedenjak** [comment]
The HTML and CSV show too little as we cannot traverse references. From the looks of it, hazelcast is consuming about half a megabyte of heap.
It is also possible that the memory has been freed after you destroy the map but hasn't yet been garbage collected. Unfortunately, this is as much as we can conclude from what you have provided.

The best way for us to analyse any memory leak would be if you could start your process with the `-XX:+HeapDumpOnOutOfMemoryError`  flag and if you could cause an OOME by repeating the process of creating a map, filling it and destroying it. You can specify the heap dump path via `-XX:HeapDumpPath`.

**Comment by vijaymaddi** [comment]
thanks for the help @mmedenjak , we will try with the options you have provided  for heap dump.

**Comment by ahmetmircik** [comment]
Hi, committed virtual memory is not the right place to determine a map leak but JProfiler can be helpful here. After map#destroy call and collection of garbage, you should see on JProfiler that final used memory same with the starting one. (Or alternatively you can see used memory programmatically with this : `Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()`)

Tried to reproduce the issue with the versions 3.8, 3.8.9 and 3.10.4, all showed no leak, used this test and JProfiler: 

```
HazelcastInstance hz = Hazelcast.newHazelcastInstance();

for (int ii = 0; ii < 5; ii++) {
    IMap map = hz.getMap("test");

    Random random = new Random();

    byte[] values = new byte[128];
    random.nextBytes(values);

    for (int i = 0; i < 10000000; i++) {
        map.set(i, values);
    }

    map.destroy();
}
```

**Comment by mmedenjak** [comment]
@vijaymaddi any updates on the issue?

**Comment by mmedenjak** [comment]
Closing as there are no updates. Please reopen once you have new information or if you encounter a new issue.
